### ISSUE TITLE
State cookie lost when Idserver receives a callback from ADFS after logout

### ISSUE THREAD

**Issue by tiwari-abhishek** [IdentityServer/IdentityServer3#3268]
- [x] I read and understood how to enable [logging](https://identityserver.github.io/Documentation/docsv2/configuration/logging.html)

### Question / Issue
I am using ADFS as an identity provider and IdentityServer is setup as a relying party in ADFS. Login works fine and the user gets redirected back to client application.

When user hits logout, I use the below code to preserve the signout message so that I can look it up after user returns from ADFS. But there no cookies on return so I get stuck. What am I missing? 

         `app.UseWsFederationAuthentication( new WsFederationAuthenticationOptions
                {
                    Wtrealm = "urn:identityServer",
                    MetadataAddress = "https://fedservice.mydomain.local/federationmetadata/2007-06/federationmetadata.xml",
                    AuthenticationType = "adfs",
                    Caption = "ADFS",
                    SignInAsAuthenticationType = signInAsType,
                    Notifications = new WsFederationAuthenticationNotifications()
                    {
                        SecurityTokenValidated = n =>
                        {
                            var filter = new NormalizingClaimsFilter();
                            var incomingClaimsFromAdfs = n.AuthenticationTicket.Identity.Claims.ToList();
                            var subject = incomingClaimsFromAdfs.Find(c => c.Type == ClaimTypes.Upn);
                            incomingClaimsFromAdfs.Add(new Claim("sub", subject.Value));
                            var normalizedClaims = filter.Filter(incomingClaimsFromAdfs);
                            var claimsIdentity = new ClaimsIdentity(normalizedClaims,
                                n.AuthenticationTicket.Identity.AuthenticationType);
                            n.AuthenticationTicket = new AuthenticationTicket(claimsIdentity,
                                n.AuthenticationTicket.Properties);
                            return Task.FromResult(0);
                        },
                        RedirectToIdentityProvider = n =>
                        {
                            if (n.ProtocolMessage.IsSignOutMessage)
                            {
                                var user = n.OwinContext.Authentication.User;
                                var signOutMessageId = n.OwinContext.Environment.GetSignOutMessageId();
                                if (signOutMessageId != null)
                                {
                                    n.OwinContext.Response.Cookies.Append("state", signOutMessageId);
                                }
                            }
                            return Task.FromResult(0);
                        }
                    }
                }
                );

           `app.Use(async` (ctx, next) =>
            {
                var qs = ctx.Request.Query;
                var wa = qs.Get("wa");
                if (wa != null)
                {
                    if (wa == "wsignoutcleanup1.0")
                    {
                        var state = ctx.Request.Cookies["state"];
                        await ctx.Environment.RenderLoggedOutViewAsync(state);
                    }
                    else
                    {
                        await next();
                    }
                }
                else
                {
                    await next();
                }
            });`
### Relevant parts of the log file

<log goes here>

**Comment by tiwari-abhishek** [comment]
I also tried the below code.

I do get the cookie the first time middleware is kicked in, and the RenderLoggedOutViewAsync(state) method does queue it up to show the logged out view(I can see that in ctx.Environment, but somehow the cookie gets lost on the way between and finally I see a 404 on the callback url and request dies.

If I comment ctx.Environment.RenderLoggedOutViewAsync method out and directly call identity server /logout?id=<state> everything works fine.

Is this acceptable or does it seem like a hack and I should only use RenderLoggedOutViewAsync?

            `app.Map("/post-logout-callback", cb =>
            {
                cb.Run(async ctx =>
                {
                    var state = ctx.Request.Cookies["state"];
                    //**await ctx.Environment.RenderLoggedOutViewAsync(state);**
                    ctx.Response.Redirect("/logout?id=" + state);
                });
            });

**Comment by brockallen** [comment]
There are two scenarios: one where the logout starts from a client of IdSvr, and another where the logout starts from a client of ADFS. This talks about the former: https://brockallen.com/2016/02/08/single-sign-out-and-identityserver3/, and this talks about the latter: https://brockallen.com/2016/02/09/federated-sign-out-and-identityserver3/

I wonder if you're mixing the two scenarios?

**Comment by tiwari-abhishek** [comment]
Hi Brock,
I am using IdentityServer3, ADFS 2016.

The objective is basically to let users Authenticate using ADFS (through identity server external login, much like Google, Facebook) but use identity server as the Open Id provider to client applications so client applications would have nothing to do with ADFS directly.

The setup is like this:
 - Identity server is the OpenId provider for client applications
- I am using UseWsFederationAuthentication middleware to configure ADFS in Identity server as mentioned in my comment above
- Identity Server is setup as a relying party in ADFS with a "post-logout-callback" URL where ADFS would redirect in case of logout
- The client application at Identity server is configured with a post logout redirect URL to do final redirect to client application once all the logout stuff is done at ADFS and Identityserver.

Login Flow:
- The client requests token from Identity Server
- Identity Server redirects to ADFS for authentication
- user logs in at ADFS and redirected back to identity server
- The UseWsFederationAuthentication.SecurityTokenValidated handler triggers and I get to grab ADFS claims, do some mappings and issue an Authentication Ticket with normalized claims, including "sub".
- I land up in a CustomUserService.AuthenticateExternalAsync(ExternalAuthenticationContext context) method where I process additional claims and set context.AuthenticateResult
- Then identity server checks and loads consent screen if needed, user agrees
- GetProfileDataAsync invoked, adds more claims as per consent
- Token signing kicks in and issues id_token and access_token and user is redirected back to the client application
- Client gets the token and can work with APIs now

Logout Flow:
- The logout is initiated by the client application registered at Identity Server
- Identity server receives the id_token_hint and creates a sign-out message
- I grab this sign-out message in a "state" cookie(I saw that in one of your posts somewhere, but it was for openid middleware.  I used the same logic inside the ws-fed middleware)
-  Identity server redirects browser to ADFS, logs user out, and comes back to identity server "post-logout-callback" endpoint which is setup as a middleware in identity server as described above
- I do a app.map as described below:
       `app.Map("/post-logout-callback", cb =>
        {
            cb.Run(async ctx =>
            {
                var state = ctx.Request.Cookies["state"];
                //**await ctx.Environment.RenderLoggedOutViewAsync(state);**
                ctx.Response.Redirect("/logout?id=" + state);
            });
        });

Now this middleware gets triggered a few times, 2-3 I guess, (I dont know if it should). I can see the "state" cookie once and then if I use 
       `await ctx.Environment.RenderLoggedOutViewAsync(state);
I do see it added to the ctx.Environment as a Queued thing for redirection. But then this middleware is kicked again couple of times and it seems to loose the "state" somewhere, because its missing when I debug on subsequent invocations.

So I just tried a plain redirect ctx.Response.Redirect("/logout?id=" + state); to the logout endpoint with the "state" parameter that I am able to grab the first time. This seems to work.

2 questions:
- Is this the right way to do it? If I don't use await ctx.Environment.RenderLoggedOutViewAsync(state); and directly call /logout?id=<state> does it skip some cleanup and introduce vulnerability? Will this in certain scenarios?

- What is the preferred way of configuring ADFS as identity provider, should it be OpenId or Ws-Fed. To my client application it doesn't really matter because I would send them Identity Server issued Tokens only.

Sorry. I know this is a long post :-)

**Comment by leastprivilege** [comment]
any update?

**Comment by tiwari-abhishek** [comment]
Hi Dominic,
I think I posted too much info, but in short my setup is like this. I got 2 idservers and an ADFS 2016 instance. The first one is the OpenId provider, which connects to another one which is ws-fed provider, which might or might not connect to ADFS depending upon what user does.

So on login and logout there are additional hops involved. While its ok for login, on logout somehow the state cookie goes missing from the middle idserver which is the ws-fed provider.

I am still using the second way, of directly calling the /logout endpoint with the "state" pulled back from the cookie. It seems to work and does log out from all involved parties including ADFS, IDserver as ws-fed provider and another one on top of all that as OpenId provider.

But I am still wondering whether it is fine to do it this way and whether it will destroy all user information instead of the `await ctx.Environment.RenderLoggedOutViewAsync(state);`
